app/helpers/ExerciseConfig/Pipeline/Box/Boxes/JudgeBox.php

```diff
@@ -113,19 +109,32 @@ private function getJudgeBinaryAndArgs(): array {
 
     $judgeType = null;
     if ($this->hasInputPortValue(self::$JUDGE_TYPE_PORT_KEY)) {
-      $judgeType = $this->getInputPortValue(self::$JUDGE_TYPE_PORT_KEY)->getValue();
+      $judgeType = strtolower($this->getInputPortValue(self::$JUDGE_TYPE_PORT_KEY)->getValue());
     }
 
-    // judge type decision logic
-    if (empty($judgeType) || strtolower($judgeType) === self::$RECODEX_NORMAL_TYPE) {
-      return [ConfigParams::$JUDGES_DIR . self::$RECODEX_NORMAL_TYPE, []];
-    } else if (strtolower($judgeType) === self::$RECODEX_SHUFFLE_TYPE) {
-      return [ConfigParams::$JUDGES_DIR . self::$RECODEX_SHUFFLE_TYPE, []];
-    } else if (strtolower($judgeType) === self::$DIFF_TYPE) {
-      return [self::$DIFF_BINARY, []];
+    static $judgeTypes = null;
+    if ($judgeTypes === null) {
+      $judgeTypes = [
+        'recodex-judge-normal' => [ConfigParams::$JUDGES_DIR . 'recodex-judge-normal', []],               // default token judge (respecting newlines)
+        'recodex-judge-float' => [ConfigParams::$JUDGES_DIR . 'recodex-judge-normal', ['-r']],            // judge comparing float values with some margin of error
+        'recodex-judge-normal-newline' => [ConfigParams::$JUDGES_DIR . 'recodex-judge-normal', ['-n']],   // default token judge (which treats \n as normal whitespace)
+        'recodex-judge-float-newline' => [ConfigParams::$JUDGES_DIR . 'recodex-judge-normal', ['-rn']],   // judge comparing float values (which treats \n as normal whitespace)
+        'recodex-judge-shuffle' => [ConfigParams::$JUDGES_DIR . 'recodex-judge-shuffle', ['-i']],         // judge ignoring order of tokens on a line
+        'recodex-judge-shuffle-rows' => [ConfigParams::$JUDGES_DIR . 'recodex-judge-shuffle', ['-r']],    // judge ignoring order of rows
+        'recodex-judge-shuffle-all' => [ConfigParams::$JUDGES_DIR . 'recodex-judge-shuffle', ['-i','-r']],// judge ignoring order of tokens on a each line and order of rows
+        'recodex-judge-shuffle-newline' => [ConfigParams::$JUDGES_DIR . 'recodex-judge-shuffle', ['-i','-n']], // judge ignoring order of tokens (which treats \n ...)
+        'diff' => ["/usr/bin/diff", []],                                                               // diff (binary-safe) judge
+      ];
     }
 
-    throw new ExerciseConfigException("Unknown judge type");
+    // judge type decision logic
+    if (empty($judgeType)) {
+      return reset($judgeTypes);
+    } else if (!empty($judgeTypes[$judgeType])) {
```

Comment:
I would probably use `array_key_exists` here - if the key is in the array, it is almost certain it will have a non-empty value.

Reply:
I would not for two reasons: array_key_exists is longer to write and is more error prone (the least thing I can do is to check that a nonempty value is really there).

Reply:
Okay, let's say it's a matter of personal taste... (array_key_exists is more explicit about what it does)